### PR TITLE
Add search option

### PR DESCRIPTION
**Pull request by orenc17**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -37,13 +36,112 @@ def buckets() -> str:
...
 @app.route("/buckets/<bucket_name>/<path:path>")
 def view_bucket(bucket_name: str, path: str) -> str:
     s3_client = boto3.client("s3", **AWS_KWARGS)
-
+    responses = []
     try:
-        response = s3_client.list_objects_v2(Bucket=bucket_name, Prefix=path, Delimiter="/")
+        marker = ''
+        while True:
+            response = s3_client.list_objects_v2(Bucket=bucket_name,
```
Comment (romanzdk) [inline app.py:137]: Hi, thanks for the PR! I tried to run your version on my machine but whenever I open any bucket I get "An error occurred
An unknown error occurred: An error occurred (InvalidArgument) when calling the ListObjectsV2 operation: The continuation token provided is incorrect".

File: app.py [hunk 2]
```diff
@@ -37,13 +36,112 @@ def buckets() -> str:
...
+
+    if search_param:
+        contents = list(filter(lambda x: search_param in x['name'], contents))
+    contents = sorted(contents, key=lambda x: x["type"], reverse=True)
+    return contents
+
+
+@app.route("/search//buckets/<bucket_name>", defaults={"path": ""})
+@app.route("/search/buckets/<bucket_name>/<path:path>")
+def search_bucket(bucket_name: str, path: str) -> str:
```
Comment (romanzdk) [inline app.py:81]: Unfortunately I was not able to test this out due to the error described below.

**Review by romanzdk — CHANGES_REQUESTED** [review]

**Comment by orenc17** [comment]
Hope my fix helps

**Comment by romanzdk** [comment]
Unfortunately that did not fix the issue. The error happens on line 139 in `view_bucket` function throwing an `botocore.exceptions.ClientError: An error occurred (InvalidArgument) when calling the ListObjectsV2 operation: The continuation token provided is incorrect` error.

**Comment by orenc17** [comment]
can you give me some test credentials? so i could test on my end

**Comment by orenc17** [comment]
Can you change the original marker value to None and check?

**Comment by romanzdk** [comment]
So `marker=None` did not work because a string is expected. I had to (ugly) hack the first request to not use the continuation token at all. Something like 

```python
if not marker:
                response = s3_client.list_objects_v2(Bucket=bucket_name,
                                                 Prefix=path,
                                                 Delimiter="/",
                                                 )
            else:    
                response = s3_client.list_objects_v2(Bucket=bucket_name,
                                                 Prefix=path,
                                                 Delimiter="/",
                                                 ContinuationToken=marker)
```

worked for me. However, when I submit the search I get the same error again. The same issue probably.

File: templates/bucket_contents.html [hunk 3]
```diff
@@ -1,4 +1,8 @@
 {% extends "base.html" %} {% block content %}
+<form action="{{ url_for('search_bucket', bucket_name=bucket_name, path=path) }}">
+  <input class="form-control" name="search" placeholder="{{ search_param }}" autocomplete="off" autofocus="autofocus" type="text">
+  <input type="submit" value="Submit" />
```
Comment (romanzdk) [inline templates/bucket_contents.html:4]:
```suggestion
  <input type="submit" value="Search" />
```

**Comment by romanzdk** [comment]
Please mention this new feature also in the README

**Comment by romanzdk** [comment]
Now it fails again on view bucket. The error is on the newly added line that converts list of dicts to set and to list again. Dict is unhashable. Please test the changes before push.

File: templates/bucket_contents.html [hunk 4]
```diff
@@ -1,4 +1,8 @@
 {% extends "base.html" %} {% block content %}
+<form action="{{ url_for('search_bucket', bucket_name=bucket_name, path=path) }}">
+  <input class="form-control" name="search" placeholder="{{ search_param }}" autocomplete="off" autofocus="autofocus" type="text">
+  <input type="submit" value="Search" />
+</form>
```
Comment (romanzdk) [inline templates/bucket_contents.html:5]:
```suggestion
<form
  action="{{ url_for('search_bucket', bucket_name=bucket_name, path=path) }}">
  <input class="form-control" name="search" placeholder="{{ search_param }}"
    autocomplete="off" autofocus="autofocus" type="text">
  <button class="btn waves-effect waves-light" type="submit" name="action">
    Search
  </button>
```

**Comment by romanzdk** [comment]
still an error on view bucket...

```
File "/Users/romanzdk/Repos/s3-browser/app.py", line 163, in view_bucket
    contents = parse_responses(responses, s3_client, bucket_name, search_param)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/romanzdk/Repos/s3-browser/app.py", line 59, in parse_responses
    S3Entry({
TypeError: S3Entry.__init__() missing 1 required positional argument: 'type'
```

**Pull request merged**